pkg/controller/mizar/mizar-service-controller.go

```diff
@@ -0,0 +1,292 @@
+type MizarServiceController struct {
+	netClientset        *arktos.Clientset
+	kubeClientset       *kubernetes.Clientset
+	serviceLister       corelisters.ServiceLister
+	netLister           arktosextv1.NetworkLister
+	serviceListerSynced cache.InformerSynced
+	syncHandler         func(eventKey KeyWithEventType) error
+	queue               workqueue.RateLimitingInterface
+	recorder            record.EventRecorder
+	grpcHost            string
+}
+
+// NewMizarServiceController starts mizar service controller
+func NewMizarServiceController(kubeClientset *kubernetes.Clientset, netClientset *arktos.Clientset, serviceInformer coreinformers.ServiceInformer, arktosInformer arktosinformer.NetworkInformer, grpcHost string) *MizarServiceController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClientset.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
+
+	c := &MizarServiceController{
+		kubeClientset:       kubeClientset,
+		netClientset:        netClientset,
+		serviceLister:       serviceInformer.Lister(),
+		netLister:           arktosInformer.Lister(),
+		serviceListerSynced: serviceInformer.Informer().HasSynced,
+		queue:               workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		recorder:            eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "mizar-service-controller"}),
+		grpcHost:            grpcHost,
+	}
+
+	serviceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.createService,
+		UpdateFunc: c.updateService,
+		DeleteFunc: c.deleteService,
+	})
+	c.serviceLister = serviceInformer.Lister()
+	c.serviceListerSynced = serviceInformer.Informer().HasSynced
+
+	c.syncHandler = c.syncService
+
+	return c
+}
+
+// Run create, update, delete service on mizar side
+func (c *MizarServiceController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	klog.Infof("Starting node controller")
+	klog.Infoln("Waiting cache to be synced.")
+	if ok := cache.WaitForCacheSync(stopCh, c.serviceListerSynced); !ok {
+		klog.Fatalln("Timeout expired during waiting for caches to sync.")
+	}
+	klog.Infoln("Starting custom controller.")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Info("shutting down node controller")
+}
+
+func (c *MizarServiceController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MizarServiceController) processNextWorkItem() bool {
+	workItem, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	eventKey := workItem.(KeyWithEventType)
+	key := eventKey.Key
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(eventKey)
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Handle %v of key %v failed with %v", "serivce", key, err))
```

Comment:
minor : fmt.Errorf("Handle %v of key %v failed with %v", "serivce", key, err -> fmt.Errorf("Handle service of key %v failed with %v", key, err